### PR TITLE
Simperfit feature/adding hal support

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #608
| License       | MIT
| Doc PR        | todo

Follow up of #608.

* [x] Fix and complete unit tests
* [ ] Entrypoint support
* [ ] Exception support (maybe in another PR)

### CODE REVIEW

File: src/Serializer/AbstractItemNormalizer.php [hunk 1]
```diff
@@ -347,12 +289,70 @@ private function getFactoryOptions(array $context) : array
...
+     * {@inheritdoc}
+     */
+    protected function getAttributeValue($object, $attribute, $format = null, array $context = [])
+    {
+        $propertyMetadata = $this->propertyMetadataFactory->create($context['resource_class'], $attribute, $this->getFactoryOptions($context));
+
+        $attributeValue = $this->propertyAccessor->getValue($object, $attribute);
+        $type = $propertyMetadata->getType();
+
+        if (
```
Comment (Simperfit) [inline src/Serializer/AbstractItemNormalizer.php]: I feel it  looks like a strange indent ?

File: src/Hal/Serializer/ItemNormalizer.php [hunk 2]
```diff
@@ -72,19 +77,153 @@ public function normalize($object, $format = null, array $context = [])
...
+    }
+
+    /**
+     * Gets the IRI of the given relation.
+     *
+     * @param array|string $rel
+     *
+     * @return string
+     */
+    private function getRelationIri($rel)
```
Comment (Simperfit) [inline src/Hal/Serializer/ItemNormalizer.php]: typehint missing

File: src/Hal/Serializer/ItemNormalizer.php [hunk 3]
```diff
@@ -72,19 +77,153 @@ public function normalize($object, $format = null, array $context = [])
...
+    /**
+     * Gets HAL components of the resource: states, links and embedded.
+     *
+     * @param object      $object
+     * @param string|null $format
+     * @param array       $context
+     *
+     * @return array
+     */
+    private function getComponents($object, $format = null, array $context)
```
Comment (Simperfit) [inline src/Hal/Serializer/ItemNormalizer.php]: typehint missing

File: src/Hal/Serializer/ItemNormalizer.php [hunk 4]
```diff
@@ -72,19 +77,153 @@ public function normalize($object, $format = null, array $context = [])
...
-        // Avoid issues with proxies if we populated the object
-        if (isset($data['_links']['self']['href']) && !isset($context['object_to_populate'])) {
-            $context['object_to_populate'] = $this->iriConverter->getItemFromIri($data['_links']['self']['href'], true);
+        throw new RuntimeException(sprintf('%s is a read-only format.', self::FORMAT));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getAttributes($object, $format = null, array $context)
```
Comment (Simperfit) [inline src/Hal/Serializer/ItemNormalizer.php]: typehint missing

File: src/Serializer/AbstractItemNormalizer.php (hunk 1 above)
Comment (dunglas) [inline src/Serializer/AbstractItemNormalizer.php]: Why? What do you suggest?

Comment (Simperfit) [inline src/Serializer/AbstractItemNormalizer.php]:
```php
if ((is_array($attributeValue) || $attributeValue instanceof \Traversable)
    &&
     ...
```

**Comment by Simperfit** [comment]
LGTM with last comments

**Pull request merged**